### ISSUE TITLE
[MENU] Revamping of Game Menus

### ISSUE THREAD

**Issue by MotoLegacy** [nzp-team/nzportable#7]
With the in-progress to move the FTE build menus to MenuQC via the [quakec/menuqc](https://github.com/nzp-team/quakec/tree/menuqc) branch in order to fix issues such as the custom map button selection (https://github.com/nzp-team/nzportable/issues/5) poor setting implementation, and improper aspect scaling, the opportunity arises to give each individual menu and submenu a coat of visual paint to become more intuitive for the user to navigate as well as generally improve upon its design.

The goal of this issue is to serve as community input and discussion for this general menu design and per-menu feedback in order to ensure a well-designed end product.

With that being said, here is my current design idea for the game's main menu:

<img src="https://user-images.githubusercontent.com/26030250/164283276-d32c9f35-1c5b-497f-bf94-ab02a4e2178b.png" width="500"/>

The background image is temporary, I am also looking for feedback on what exactly to populate the background with instead, whether that be a still image like before or a rendered demo like the NZ:P Unofficial Patch had done in the past.

This issue will be updated to reflect community approved designs and the current submenu needing feedback and approval, so refer back to this primary comment often and be sure to be explicit about which menu you're providing feedback for.

Thanks,
Moto

**Comment by DerpedCrusader** [comment]
rendered backgrounds are cool, but loading times can be PSP  kinda brutal.

**Comment by MotoLegacy** [comment]
> rendered backgrounds are cool, but loading times can be PSP  kinda brutal.

This was offset in the Unofficial Patch by preloading anything used into the heap during initialization, as well as just generally not having to go through all of the spawn functions, so general full map load time doesn't reflect the theoretical load time for the still demo playback.

**Comment by MotoLegacy** [comment]
Blubs, Naievil, and I have come up with a concept to where the menu background would be a randomly selected map, the camera being either at the player's normal spawn position, or one/several camera entities placed into the map. These cameras would also prove useful for adding support for panning through the level during the Game Over screen. Looking for feedback.

**Comment by DerpedCrusader** [comment]
kinda like the Half life 2/Source engine game main menus, I'd love that personally.

and the two birds with 1 stone is great too with the game over screen camera

**Comment by MotoLegacy** [comment]
> kinda like the Half life 2/Source engine game main menus, I'd love that personally.
> 
> and the two birds with 1 stone is great too with the game over screen camera

Sort of. But with more motion, since the mapper will likely have lots of fly-throughs like CoD does typically. Still looking for input from other users.

**Comment by Doc-Cynthetic** [comment]
> Blubs, Naievil, and I have come up with a concept to where the menu background would be a randomly selected map, the camera being either at the player's normal spawn position, or one/several camera entities placed into the map. These cameras would also prove useful for adding support for panning through the level during the Game Over screen. Looking for feedback.

I think this is overall the best solution, that way we get different images based on the installed maps. Do you think it would be appropriate to add a method to load custom background images/.gifs, this way players could choose to set a certain background, or on lower end devices the player could simply choose a static image over the rendered background

**Comment by MotoLegacy** [comment]
> [...] Do you think it would be appropriate to add a method to load custom background images/.gifs [...]

Yes I think this would be a good option to have, at the very least as a cvar in the configuration file. I will plan for this system in the design but it might be hard to account for it as an actual setting in the menu. Let's discuss that when we get to the settings page.

**Comment by Doc-Cynthetic** [comment]
> I will plan for this system in the design but it might be hard to account for it as an actual setting in the menu. 
>

I think for intents and purposes it doesn't have to be in the settings on its own, but nonetheless useful for customization. But i like the button layout you chose for the example image, However i disagree with the usage of all-caps for the buttons, i think perhaps the buttons should have some sort of background "image" rather than being slightly greyed when its selected (ex. some blood spatter or zombie eyes) Just to give extra color and personality to the menu selection. Perhaps the process of integrating the "video/images" could be a random selection from a folder rather than forcing the game to render on the main menu. (A map maker could simply make a recording of their camera shots and simply have the video in the game files, not sure if you can get videos playing in the back of the main menu, but i figure this might be easier than trying to render those shots on the fly)

**Comment by MotoLegacy** [comment]
After the discussion that occurred yesterday I decided to create a PSP-focused mockup with a few general changes, as well as provide a photo of the mockup being displayed on the PSP screen itself to help capture our resolution limitation.

General changes:
* I've added a dark-red border on the three visible sides of the highlight band to try and give it more visual character.
* Text is now right-aligned to remain better enhance consistency (it is better to have deadspace on the left side of the band as opposed to the right as the right edge is the most visible).
* The mockup displays placeholder/intended maximum character count text as opposed to the menu text, to better focus on general design.
  * 24 characters for menu items
  * 48 characters for item descriptions

![menuv2](https://user-images.githubusercontent.com/26030250/164914337-121b1615-a04e-452e-ac45-103253b5095e.png)

![F955568C-47FC-4C50-9932-32FF38DD690B](https://user-images.githubusercontent.com/26030250/164914347-bebe0ad8-acd2-4e23-a167-4ebc2ee44dec.jpeg)

Let me know what everyone thinks.

**Comment by DerpedCrusader** [comment]
The right aligned sounds nice, but i would love to see it with non-placeholder text

The border around selection looks nice

Is there any way to have it gradient into transparency on the left side, kinda like the MW2/3 menu does?

**Comment by MotoLegacy** [comment]
> i would love to see it with non-placeholder text
This request undermines the reason the placeholder text is there in the first place (to focus on general design critique since that is what the conversation had become), though I submitted to this request anyway.

Because I had done this, I created 2 variants for the request. One is with the same ("static") width of elements, and another where the width is dependent ("dynamic") on the maximum size of the currently displayed elements.

There is a discussion to be had about which to side with (it becomes frustrating because while the dynamic option is more efficient in terms of real-estate it brings more design inconsistencies, which the point was to avoid. Would look rather jarring.) so I look forward to hearing suggestions.

I added the texts "NACHT DER UNTOTEN" (17 characters) and "SUPER MARIO 64 ZOMBIES" (22 characters) to the static mock-up in case an argument is going to try to be made about just not expecting/accounting for 20+ characters and shrinking the static width.

![menuv2-main-static](https://user-images.githubusercontent.com/26030250/164934659-ed8383dc-cb60-4dd1-8718-8dee27f8fba2.png)

![menuv2-main-dynamic](https://user-images.githubusercontent.com/26030250/164934662-a949d7cf-a5ad-4616-8af0-bcbd6189df3f.png)

> Is there any way to have it gradient into transparency on the left side, kinda like the MW2/3 menu does?

In a *very* inefficient way, yes, a feathering effect can be done by just filling X relative pixels with a slightly less opaque rectangle although this is very taxing and also just a bad means of implementation. An image could be used and stretched, but that feels like a total waste of memory and cannot be controlled as smoothly for differing aspect ratios.

I feel as though not much is gained/lost anyway, though I have made an additional mock-up for it.

![menuv2-feathered](https://user-images.githubusercontent.com/26030250/164939299-d8fbf993-c960-48ee-bc74-fd323de68c0a.png)

**Comment by Doc-Cynthetic** [comment]
I REALLY appreciate the implications of the first image, Moto. I think the layout and design in that image is absolutely spectacular, if it could be integrated to show 2-3 of the most recently played maps or w/e where you put ndu and “Super Mario 64 Zombies” (xD) would be a really cool and functional addition to the main screen. Considering the main focus for the PSP, I think the increase of menu-to-gameplay fluidity plays into exactly what the PSP represents; pick up the game, just in your favorite map, play for 15 minutes, back to business.

I think the gradient could go either way, it looks really really clean, but if it takes too much energy to produce then it may not be worth the effort (depending on the amount of processing it actually takes)

personally I think it’s looking really really nice so far, the border color change is very nice to look at :)

**Comment by MotoLegacy** [comment]
Thanks for the input, @Doc-Cynthetic. I do not think the feathered edges are worth the effort in GPU calls or implementation but I'm glad they look nice in concept at least.

That "Jump Back In" note was absolutely fantastic, that is definitely something I will include in the actual menu. Something to get right into the lobby setting with recently loaded maps.

I need an opinion on the static width though, do you think all of the empty space on the left side of entries is off-putting? It would 100% need to be static if the Jump-In feature is added.

**Comment by Doc-Cynthetic** [comment]
I think it looks perfect as-is. As much as my brain wants to ask for left-justified text, it will look cleaner justified right and easier read on the PSP. I think the static width works very pleasantly with the UI. 

Would it be difficult to mock-up a “left-justified” version where the border stops at the end of the characters on the right side? This will reduce the look of the menu “box” but I think the different sized borders would free up some extra background space for its eventual application.

with the current layout, it would be easy (and look pretty great) to use a custom background where the menu selection boxes are encased/incorporated in the background image. So I see advantages and drawbacks for both, give me some time to make a background image like this and I can show you a better example.

**Comment by MotoLegacy** [comment]
My concern with the idea of having the width of the band cut at the end of the text with it being left-align is the inconsistency of it — not only is it visually unappealing but it also can be jarring on devices with mice input for determining where the actual selection box is.

**Comment by Paindeluxe** [comment]
When would this be finished and given that there is a help wanted tag how could we help?
